### PR TITLE
feat(component): add save annotation modal

### LINKED ISSUE #1514: [refactor]: update the version of white-snapshot to remove html2canvas support.
Maybe I could just update the version of white-snapshot to remove html2canvas support.

_Originally posted by @hyrious in https://github.com/netless-io/flat/pull/1512#discussion_r866543005_

### PR DESCRIPTION
**Pull request by hyrious**
closes #1514

### CODE REVIEW

File: packages/flat-components/src/components/SaveAnnotationModal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+export const SaveAnnotationModal: React.FC<SaveAnnotationModalProps> = ({
+    visible,
+    onClose,
+    images,
+}) => {
+    const { t } = useTranslation();
+
+    const params = useMemo(() => {
+        const limit = pLimit(1);
+        return images.map(image => limit(image));
```
Comment (BlackHole1) [inline packages/flat-components/src/components/SaveAnnotationModal/index.tsx:43]:
```suggestion
        return images.map(limit);
```

File: packages/flat-components/src/components/SaveAnnotationModal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+export const SaveAnnotationModal: React.FC<SaveAnnotationModalProps> = ({
+    visible,
+    onClose,
+    images,
+}) => {
+    const { t } = useTranslation();
+
+    const params = useMemo(() => {
+        const limit = pLimit(1);
```
Comment (BlackHole1) [inline packages/flat-components/src/components/SaveAnnotationModal/index.tsx:42]: Why not use `Promise.race` ?

File: packages/flat-components/src/components/SaveAnnotationModal/index.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        };
+    }, [image]);
+
+    useEffect(() => {
+        const div = ref.current;
+        if (div && canvas) {
+            div.appendChild(canvas);
+        } else if (div && div.firstChild) {
+            div.removeChild(div.firstChild);
+        }
```
Comment (BlackHole1) [inline packages/flat-components/src/components/SaveAnnotationModal/index.tsx]:
```suggestion
        if (!idiv) {
            return;
        }
        
        if (canvas) {
            div.appendChild(canvas);
        } else if (div.firstChild) {
            div.removeChild(div.firstChild);
        }
```

File: packages/flat-components/src/components/SaveAnnotationModal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        let isMounted = true;
+
+        image.then(canvas => {
+            isMounted && setCanvas(canvas);
+        });
+
+        return () => {
+            isMounted = false;
+        };
+    }, [image]);
```
Comment (BlackHole1) [inline packages/flat-components/src/components/SaveAnnotationModal/index.tsx]: use `useSafePromise` Custrom hooks

File: packages/flat-components/src/components/SaveAnnotationModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    failText = "Save image failed",
+): void {
+    try {
+        const a = document.createElement("a");
+        a.download = filename;
+        a.href = canvas.toDataURL();
+        a.style.display = "none";
+        document.body.appendChild(a);
+        a.click();
+        document.body.removeChild(a);
```
Comment (BlackHole1) [inline packages/flat-components/src/components/SaveAnnotationModal/index.tsx]:
```suggestion
        a.click();
```

No need `appendChild`

File: package.json [hunk 6]
```diff
@@ -50,6 +51,7 @@
     "eslint-plugin-react": "^7.29.4",
     "eslint-webpack-plugin": "^3.1.1",
     "fork-ts-checker-webpack-plugin": "^7.2.1",
+    "html2canvas": "^1.4.1",
```
Comment (BlackHole1) [inline package.json]: Not used. Please remove

File: desktop/renderer-app/src/utils/toolbar-apps.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import countdownSVG from "../assets/image/tool-countdown.svg";
+import saveSVG from "../assets/image/tool-save.svg";
+
+import { TFunction } from "react-i18next";
+import { apps, FastboardApp } from "@netless/fastboard-react";
+import { noop } from "lodash-es";
+import { i18n } from "./i18n";
+
+export interface RefreshAppsParams {
+    t?: TFunction;
```
Comment (BlackHole1) [inline desktop/renderer-app/src/utils/toolbar-apps.ts]: This is required, not optional

File: desktop/renderer-app/src/stores/whiteboard-store.ts [hunk 8]
```diff
@@ -615,4 +616,26 @@ export class WhiteboardStore {
...
+                // Because manager hacks room.fillSceneSnapshot, we need to hack it back.
+                const room = {
+                    state: manager,
+                    fillSceneSnapshot: manager.mainView.fillSceneSnapshot.bind(manager.mainView),
+                } as any;
+                return () =>
+                    snapshot(room, {
+                        scenePath: dir + scene.name,
+                        // TODO: configure CDN to support saving images
+                        // crossorigin: true,
```
Comment (BlackHole1) [inline desktop/renderer-app/src/stores/whiteboard-store.ts]:
```suggestion
                        // TODO: configure CDN to support saving images. @hyrious 
                        // crossorigin: true,
```

File: packages/flat-i18n/locales/zh-CN.json [hunk 9]
```diff
@@ -435,5 +435,14 @@
...
+  "time": "时间",
+  "save": "保存",
+  "save-annotation": "保存板书",
+  "tool": {
+    "monaco": "编辑器",
+    "geogebra": "几何代数",
+    "countdown": "计时器",
+    "save": "保存板书"
+  },
+  "save-annotation-failed": "板书保存失败"
```
Comment (BlackHole1) [inline packages/flat-i18n/locales/zh-CN.json]: Perhaps use the `annotation` namespace? .eg

```json
"annotation": {
  "time": "",
  "save": "保存",
  "save-action": "保持板书",
  "save-failed": "板书保存失败",
  "tools": {
    "..."
  }
}
```

File: web/flat-web/src/stores/whiteboard-store.ts [hunk 10]
```diff
@@ -585,4 +586,26 @@ export class WhiteboardStore {
...
+                const dir = manager.mainViewSceneDir;
+                // Because manager hacks room.fillSceneSnapshot, we need to hack it back.
+                const room = {
+                    state: manager,
+                    fillSceneSnapshot: manager.mainView.fillSceneSnapshot.bind(manager.mainView),
+                } as any;
+                return () =>
+                    snapshot(room, {
+                        scenePath: dir + scene.name,
+                        // TODO: configure CDN to support saving images
```
Comment (BlackHole1) [inline web/flat-web/src/stores/whiteboard-store.ts]:
```suggestion
                        // TODO: configure CDN to support saving images @hyrious 
```

File: web/flat-web/src/utils/toolbar-apps.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import countdownSVG from "../assets/image/tool-countdown.svg";
+import saveSVG from "../assets/image/tool-save.svg";
+
+import { TFunction } from "react-i18next";
+import { apps, FastboardApp } from "@netless/fastboard-react";
+import { noop } from "lodash-es";
+import { i18n } from "./i18n";
+
+export interface RefreshAppsParams {
+    t?: TFunction;
```
Comment (BlackHole1) [inline web/flat-web/src/utils/toolbar-apps.ts]:
```suggestion
    t: TFunction;
```

Above

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: packages/flat-components/src/components/SaveAnnotationModal/index.tsx (hunk 1 above)
Comment (hyrious) [reply packages/flat-components/src/components/SaveAnnotationModal/index.tsx:43]: `limit` has such type signature:

```ts
limit(fn, ...args)
```

I don't think it is right to pass map arguments (index, array) to the function.

Comment (BlackHole1) [reply packages/flat-components/src/components/SaveAnnotationModal/index.tsx:43]: Okey. You are right.

File: package.json (hunk 6 above)
Comment (hyrious) [reply package.json]: Not used, but vite dev server requires it to be found&hellip; It will show error if not found this module.

Comment (BlackHole1) [reply package.json]: Maybe we need to use: `peerDependencies`?

File: packages/flat-i18n/locales/zh-CN.json (hunk 9 above)
Comment (hyrious) [reply packages/flat-i18n/locales/zh-CN.json]: no `time`

Comment (BlackHole1) [reply packages/flat-i18n/locales/zh-CN.json]: This is just an example. 😶‍🌫️

File: packages/flat-components/src/components/SaveAnnotationModal/index.tsx (hunk 2 above)
Comment (hyrious) [reply packages/flat-components/src/components/SaveAnnotationModal/index.tsx:42]: It has different purpose than `race`.

Here the `snapshot` function is heavy, we don't want to run it very quickly and suddenly eat all CPU. So we are using `p-limit` to limit the calling of snapshot to occur slowly.

Comment (BlackHole1) [reply packages/flat-components/src/components/SaveAnnotationModal/index.tsx:42]: Okey

File: package.json (hunk 6 above)
Comment (hyrious) [reply package.json]: Maybe I could just update the version of white-snapshot to remove html2canvas support.

File: desktop/renderer-app/src/utils/toolbar-apps.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    onSaveAnnotation?: (app: FastboardApp) => void;
+}
+
+export const refreshApps = ({ t, onSaveAnnotation }: RefreshAppsParams): void => {
+    apps.clear();
+    apps.push(
+        {
+            kind: "Monaco",
+            icon: monacoSVG,
+            label: t?.("tool.monaco") || "Code Editor",
```
Comment (BlackHole1) [inline desktop/renderer-app/src/utils/toolbar-apps.ts]:
```suggestion
            label: t.("tool.monaco")
```

File: web/flat-web/src/utils/toolbar-apps.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    onSaveAnnotation?: (app: FastboardApp) => void;
+}
+
+export const refreshApps = ({ t, onSaveAnnotation }: RefreshAppsParams): void => {
+    apps.clear();
+    apps.push(
+        {
+            kind: "Monaco",
+            icon: monacoSVG,
+            label: t?.("tool.monaco") || "Code Editor",
```
Comment (BlackHole1) [inline web/flat-web/src/utils/toolbar-apps.ts]:
```suggestion
            label: t.("tool.monaco"),
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: packages/flat-i18n/locales/zh-CN.json (hunk 9 above)
Comment (BlackHole1) [reply packages/flat-i18n/locales/zh-CN.json]: @hyrious

File: package.json (hunk 6 above)
Comment (hyrious) [reply package.json]: done.

**Review by BlackHole1 — DISMISSED** [review]

**Review by crimx — APPROVED** [review]

**Pull request merged**